nighthawk/source/common/utility.cc

```diff
@@ -29,26 +29,73 @@ uint32_t determineCpuCoresWithAffinity() {
 
 } // namespace PlatformUtils
 
-Uri::Uri(const std::string& uri) : scheme_("http") {
+size_t Uri::findPortSeparatorInAuthority(absl::string_view authority) {
+  size_t colon_index = std::string::npos;
+  bool in_ipv6_address = false;
+  bool last_seen_dot = true;
+
+  // TODO(oschaaf): We need to revisit this and improve. This performs some
+  // limited hostname validation.
+  for (size_t i = 0; i < authority.size(); i++) {
+    char c = authority[i];
+    if (i == 0 && c == '[') {
+      in_ipv6_address = true;
+    } else if (in_ipv6_address && c == ']') {
+      in_ipv6_address = false;
+    } else {
+      if (!in_ipv6_address && i > 0 && c == ':') {
+        colon_index = i;
+        break;
+      } else if (c == '.') {
+        if (last_seen_dot) {
+          throw InvalidHostException("Invalid hostname");
+        }
+        last_seen_dot = true;
+        continue;
+      } else {
+        last_seen_dot = false;
```

Comment:
I reckon we probably don't need to do any of this validation; it's the user's responsibility to provide something valid, I think we only care about the `:` separator. Otherwise, we get in the business of doing what other libraries (e.g. Ares) are already doing. This then simplifies quite a bit.